src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -5311,6 +5313,365 @@ class StubGenerator: public StubCodeGenerator {
+    }
+  }
+
+  // result = r0 - return value. Contains initial hashcode value on entry.
+  // ary = r1 - array address
+  // cnt = r2 - elements count
+  // Clobbers: v0-v13, rscratch1, rscratch2
+  address generate_large_arrays_hashcode(BasicType eltype) {
+    const Register result = r0, ary = r1, cnt = r2;
+    const FloatRegister vdata0 = v3, vdata1 = v2, vdata2 = v1, vdata3 = v0;
+    const FloatRegister vhalf0 = v13, vhalf1 = v12, vhalf2 = v11, vhalf3 = v10;
+    const FloatRegister vmul0 = v4, vmul1 = v5, vmul2 = v6, vmul3 = v7;
+    const FloatRegister vpow = v8;  // powers of 31: <31^3, ..., 31^0>
+    const FloatRegister vpowm = v9;
+
+    assert_different_registers(ary, cnt, result);
+    assert_different_registers(vdata0, vdata1, vdata2, vdata3, vhalf0, vhalf1, vhalf2, vhalf3,
+                               vmul0, vmul1, vmul2, vmul3, vpow, vpowm);
+
+    Label SMALL_LOOP, LARGE_LOOP_PREHEADER, LARGE_LOOP, TAIL, TAIL_SHORTCUT, RELATIVE;
+
+    // Vectorization factor
+    const size_t vf = eltype == T_BOOLEAN || eltype == T_BYTE ? 8
+                      : eltype == T_CHAR || eltype == T_SHORT ? 8
+                      : eltype == T_INT                       ? 4
+                                                              : 0;
+    guarantee(vf, "unsupported eltype");
+
+    // Unroll factor
+    const size_t uf = 4;
+
+    // Effective vectorization factor
+    const size_t evf = vf * uf;
+
+    __ align(CodeEntryAlignment);
+
+    const char *mark_name = "";
+    switch (eltype) {
+    case T_BOOLEAN:
+      mark_name = "_large_arrays_hashcode_boolean";
+      break;
+    case T_BYTE:
+      mark_name = "_large_arrays_hashcode_byte";
+      break;
+    case T_CHAR:
+      mark_name = "_large_arrays_hashcode_char";
+      break;
+    case T_SHORT:
+      mark_name = "_large_arrays_hashcode_short";
+      break;
+    case T_INT:
+      mark_name = "_large_arrays_hashcode_int";
+      break;
+    default:
+      mark_name = "_large_arrays_hashcode_incorrect_type";
+      __ should_not_reach_here();
+    };
+
+    StubCodeMark mark(this, "StubRoutines", mark_name);
+
+    address entry = __ pc();
+    __ enter();
+
+    Assembler::SIMD_Arrangement load_arrangement =
+        eltype == T_BOOLEAN || eltype == T_BYTE ? Assembler::T8B
+        : eltype == T_CHAR || eltype == T_SHORT ? Assembler::T8H
+        : eltype == T_INT                       ? Assembler::T4S
+                                                : Assembler::INVALID_ARRANGEMENT;
+    guarantee(load_arrangement != Assembler::INVALID_ARRANGEMENT, "invalid arrangement");
+
+    const int multiply_by_halves =
+        load_arrangement == Assembler::T4S || load_arrangement == Assembler::T4H   ? false
+        : load_arrangement == Assembler::T8B || load_arrangement == Assembler::T8H ? true
+                                                                                   : -1;
+    guarantee(multiply_by_halves != -1, "unknown multiplication algorithm");
+
```

Comment:
It's hard to follow this logic. Try something like this:
```suggestion

    Assembler::SIMD_Arrangement load_arrangement;
    switch (eltype) {
      case T_BOOLEAN:
      case T_BYTE:
        load_arrangement = T8B;
        multiply_by_halves = true;
        break;
      case T_CHAR:
      case T_SHORT:
        load_arrangement = T8H;
        multiply_by_halves = true;
        break;
      case T_INT:
        load_arrangement = T4S;
        multiply_by_halves = false;
        break;
      default:
        ShouldNotReachHere();
    }

```

Reply:
The current implementation reflects that the decision to process a register by halves depends on the arrangement used. In the previous version of this PR, we tested for `load_arrangement` in places where `multiply_by_halves` is tested now. This way, for example, changing the arrangement for `T_CHAR`/`T_SHORT` from `T8H` to `T4H` requires only changing the arrangement itself. Using the logic you suggest would require one to be aware of the connection between `load_arrangement` and `multiply_by_halves` that must be maintained. Therefore, I recommend leaving the code as it is.

Reply:
Fixed by https://github.com/openjdk/jdk/pull/18487/commits/a824a74263ce9309fca50a90b3d0f09e9b56a5c4 , please check.